src/bio/bio_config.c

```diff
@@ -0,0 +1,532 @@
+};
+
+static int
+cap_string(const struct spdk_json_val *val, void *out)
+{
+	const struct spdk_json_val **vptr = out;
+
+	if (val->type != SPDK_JSON_VAL_STRING) {
+		return -DER_INVAL;
+	}
+
+	*vptr = val;
+	return 0;
+}
+
+static int
+cap_object(const struct spdk_json_val *val, void *out)
+{
+	const struct spdk_json_val **vptr = out;
+
+	if (val->type != SPDK_JSON_VAL_OBJECT_BEGIN) {
+		return -DER_INVAL;
+	}
+
+	*vptr = val;
+	return 0;
+}
+
+
+static int
+cap_array_or_null(const struct spdk_json_val *val, void *out)
+{
+	const struct spdk_json_val **vptr = out;
+
+	if (val->type != SPDK_JSON_VAL_ARRAY_BEGIN && val->type != SPDK_JSON_VAL_NULL) {
+		return -DER_INVAL;
+	}
+
+	*vptr = val;
+	return 0;
+}
+
+static struct spdk_json_val *
+json_value(struct spdk_json_val *key)
+{
+	return key->type == SPDK_JSON_VAL_NAME ? key + 1 : NULL;
+}
+
+static struct spdk_json_object_decoder
+subsystem_decoders[] = {
+	{"subsystem", offsetof(struct json_config_ctx, subsystem_name), cap_string},
+	{"config", offsetof(struct json_config_ctx, config), cap_array_or_null}
+};
+
+struct
+config_entry {
+	char			*method;
+	struct spdk_json_val	*params;
+};
+
+static struct spdk_json_object_decoder
+jsonrpc_cmd_decoders[] = {
+	{"method", offsetof(struct config_entry, method), spdk_json_decode_string},
+	{"params", offsetof(struct config_entry, params), cap_object, true}
+};
+
+static int
+is_addr_in_allowlist(char *pci_addr, const struct spdk_pci_addr *allowlist,
+		     int num_allowlist_devices)
+{
+	int			i;
+	struct spdk_pci_addr    tmp;
+
+	if (spdk_pci_addr_parse(&tmp, pci_addr) != 0) {
+		D_ERROR("invalid transport address %s", pci_addr);
+		return -DER_INVAL;
+	}
+
+	for (i = 0; i < num_allowlist_devices; i++) {
+		if (spdk_pci_addr_compare(&tmp, &allowlist[i]) == 0) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * Convert a transport id in the BDF form of "5d0505:01:00.0" or something
+ * similar to the VMD address in the form of "0000:5d:05.5" that can be parsed
+ * by DPDK.
+ *
+ * \param dst String to be populated as output.
+ * \param src Input bdf.
+ */
+static int
+traddr_to_vmd(char *dst, const char *src)
+{
+	char traddr_tmp[SPDK_NVMF_TRADDR_MAX_LEN + 1];
+	char vmd_addr[SPDK_NVMF_TRADDR_MAX_LEN + 1] = "0000:";
+	char *ptr;
+	const char ch = ':';
+	char addr_split[3];
+	int position, iteration;
+	int n;
+
+	n = snprintf(traddr_tmp, SPDK_NVMF_TRADDR_MAX_LEN, "%s", src);
+	if (n < 0 || n > SPDK_NVMF_TRADDR_MAX_LEN) {
+		D_ERROR("snprintf failed");
+		return -DER_INVAL;
+	}
+
+	/* Only the first chunk of data from the traddr is useful */
+	ptr = strchr(traddr_tmp, ch);
+	if (ptr == NULL) {
+		D_ERROR("Transport id not valid");
+		return -DER_INVAL;
+	}
+	position = ptr - traddr_tmp;
+	traddr_tmp[position] = '\0';
+
+	ptr = traddr_tmp;
+	iteration = 0;
+	while (*ptr != '\0') {
+		n = snprintf(addr_split, sizeof(addr_split), "%s", ptr);
+		if (n < 0) {
+			D_ERROR("snprintf failed");
+			return -DER_INVAL;
+		}
+		strcat(vmd_addr, addr_split);
```

Comment:
Should use strncat instead of strcat, for buffer safety.

I expect this isn't called often, so efficiency is probably not a big deal. But strcat/strncat is pretty inefficient, since it has to walk the length of the destination string each time. It might be more straightforward to walk a pointer forward in the buffer as we copy substrings in. I'm wondering if representing this loop as a state machine might make it easier to follow as well.

Reply:
I have made the first change but I'm going to defer the second suggestion to @sydneyvanda as the original author of this function, as this has been in use and working I would also like to make any changes in a follow-up PR if possible.

Reply:
That's fair. I'm fine with using strncat for the time being.